File: Backend/app/modules/returns/schemas/returns.py
"""
Pydantic schemas for Returns endpoints.
"""
from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.modules.orders.models import OrderPlatform, OrderFulfillmentChannel
from app.modules.returns.models import ReturnNormalizedStatus, ReturnZohoSyncStatus


class ReturnItemDetail(BaseModel):
    id: int
    linked_order_item_id: Optional[int] = None
    external_item_id: Optional[str] = None
    external_sku: Optional[str] = None
    item_name: str
    ordered_qty: int
    returned_qty: int
    cancelled_qty: int
    refunded_amount: Decimal
    item_payload: Optional[dict] = None

    model_config = ConfigDict(from_attributes=True)


class ReturnRecordBrief(BaseModel):
    id: int
    platform: OrderPlatform
    source: str
    external_record_key: str
    external_order_id: str
    external_return_id: Optional[str] = None
    linked_order_id: Optional[int] = None
    customer_name: Optional[str] = None
    customer_email: Optional[str] = None
    ordered_at: Optional[datetime] = None
    event_at: Optional[datetime] = None
    last_source_updated_at: Optional[datetime] = None
    normalized_status: ReturnNormalizedStatus
    source_status: Optional[str] = None
    source_substatus: Optional[str] = None
    reason: Optional[str] = None
    fulfillment_channel: OrderFulfillmentChannel
    order_total_amount: Decimal
    refunded_amount: Decimal
    currency: str
    zoho_salesreturn_id: Optional[str] = None
    zoho_salesreturn_number: Optional[str] = None
    zoho_sync_status: ReturnZohoSyncStatus = ReturnZohoSyncStatus.PENDING
    zoho_sync_error: Optional[str] = None
    zoho_synced_at: Optional[datetime] = None
    item_count: int = 0
    returned_qty_total: int = 0
    cancelled_qty_total: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ReturnRecordDetail(ReturnRecordBrief):
    raw_payload: Optional[dict] = None
    items: list[ReturnItemDetail] = Field(default_factory=list)


class ReturnListResponse(BaseModel):
    total: int
    skip: int
    limit: int
    items: list[ReturnRecordBrief]
    summary_counts: dict[str, int] = Field(default_factory=dict)

File: frontend/src/types/returns.ts
export type ReturnPlatform =
  | 'AMAZON'
  | 'EBAY_MEKONG'
  | 'EBAY_USAV'
  | 'EBAY_DRAGON'
  | 'ECWID'
  | 'SHOPIFY'
  | 'WALMART'
  | 'ZOHO'
  | 'MANUAL'

export type ReturnNormalizedStatus =
  | 'RETURNED'
  | 'PARTIALLY_RETURNED'
  | 'REFUNDED'
  | 'PARTIALLY_REFUNDED'
  | 'CANCELLED'
  | 'PARTIALLY_CANCELLED'
  | 'UNKNOWN'

export type ReturnIntegrationSyncStatus = 'IDLE' | 'SYNCING' | 'ERROR'

export type ReturnZohoSyncStatus =
  | 'PENDING'
  | 'READY_TO_SYNC'
  | 'MISSING_LOCAL_ORDER'
  | 'MISSING_ZOHO_ORDER'
  | 'MISSING_LINE_ITEM_MAPPING'
  | 'QUANTITY_CONFLICT'
  | 'ALREADY_SYNCED'
  | 'SYNCED'
  | 'ERROR'

export interface ReturnItemDetail {
  id: number
  linked_order_item_id: number | null
  external_item_id: string | null
  external_sku: string | null
  item_name: string
  ordered_qty: number
  returned_qty: number
  cancelled_qty: number
  refunded_amount: string
  item_payload: Record<string, unknown> | null
}

export interface ReturnRecordBrief {
  id: number
  platform: ReturnPlatform
  source: string
  external_record_key: string
  external_order_id: string
  external_return_id: string | null
  linked_order_id: number | null
  customer_name: string | null
  customer_email: string | null
  ordered_at: string | null
  event_at: string | null
  last_source_updated_at: string | null
  normalized_status: ReturnNormalizedStatus
  source_status: string | null
  source_substatus: string | null
  reason: string | null
  order_total_amount: string
  refunded_amount: string
  currency: string
  item_count: number
  returned_qty_total: number
  cancelled_qty_total: number
  zoho_salesreturn_id: string | null
  zoho_salesreturn_number: string | null
  zoho_sync_status: ReturnZohoSyncStatus
  zoho_sync_error: string | null
  zoho_synced_at: string | null
  created_at: string
  updated_at: string
}

export interface ReturnRecordDetail extends ReturnRecordBrief {
  raw_payload: Record<string, unknown> | null
  items: ReturnItemDetail[]
}

export interface ReturnListResponse {
  total: number
  skip: number
  limit: number
  items: ReturnRecordBrief[]
  summary_counts: Record<string, number>
}

export interface ReturnSyncStateResponse {
  id: number
  platform_name: string
  last_successful_sync: string | null
  current_status: ReturnIntegrationSyncStatus
  last_error_message: string | null
  updated_at: string
}

export interface ReturnSyncStatusResponse {
  platforms: ReturnSyncStateResponse[]
  total_records: number
  counts_by_status: Record<string, number>
}

export interface ReturnSyncRequest {
  platform?: string
}

export interface ReturnSyncRangeRequest {
  platform?: string
  since: string
  until: string
}

export interface ReturnSyncResponse {
  platform: string
  new_records: number
  updated_records: number
  new_items: number
  linked_orders: number
  linked_items: number
  skipped_duplicates: number
  errors: string[]
  success: boolean
}

export interface ReturnZohoLineValidationResponse {
  return_item_id: number
  linked_order_item_id: number | null
  quantity: number
  zoho_item_id: string | null
  zoho_salesorder_item_id: string | null
  status: ReturnZohoSyncStatus
  message: string | null
}

export interface ReturnZohoValidationResponse {
  record_id: number
  status: ReturnZohoSyncStatus
  blockers: string[]
  zoho_salesorder_id: string | null
  zoho_salesreturn_id: string | null
  zoho_salesreturn_number: string | null
  line_items: ReturnZohoLineValidationResponse[]
}
